File: packages/repo-gc-python/python/repo_gc_python/heuristics/error_swallow.py
"""Empty error handler detection -- #1 LLM-generated anti-pattern (CATCHALL paper)."""

from ..cli import Threshold
from ..discovery import PythonFile
from ..parsing import FileInfo
from ..types import Finding, FindingKind, Severity


def analyze(
    pfile: PythonFile,
    info: FileInfo,
    threshold: Threshold,
    counter: int,
) -> Finding | None:
    limit = threshold.empty_catch_limit
    if info.empty_catch_count <= limit:
        return None

    if info.empty_catch_count >= 5:
        severity = Severity.High
    elif info.empty_catch_count >= 3:
        severity = Severity.Medium
    else:
        severity = Severity.Low

    return Finding(
        id=f"es-{counter:03d}",
        kind=FindingKind.ErrorSwallow,
        severity=severity,
        confidence=0.75,
        path=str(pfile.relative_path),
        summary="",
        reasons=[],
        evidence=[
            f"empty_catch_count: {info.empty_catch_count}",
            f"limit: {limit}",
        ],
        suggested_next_step="",
        estimated_tokens=None,
    )

File: packages/repo-gc/src/shared/heuristics/error-swallow.ts
import { Finding, FindingKind, Severity, type FileData } from '../types';
import { type Thresholds } from '../threshold';

export function analyzeErrorSwallow(
  data: FileData,
  thresholds: Thresholds,
  idCounter: { value: number },
): Finding | null {
  const count = data.emptyCatchCount;
  if (count <= thresholds.emptyCatchLimit) return null;

  let severity: Severity;
  if (count >= 5) {
    severity = Severity.High;
  } else if (count >= 3) {
    severity = Severity.Medium;
  } else {
    severity = Severity.Low;
  }

  return {
    id: `es-${String(idCounter.value++).padStart(3, '0')}`,
    kind: FindingKind.ErrorSwallow,
    severity,
    confidence: 0.75,
    path: data.relativePath,
    summary: '',
    reasons: [],
    evidence: [
      `empty_catch_count: ${count}`,
      `limit: ${thresholds.emptyCatchLimit}`,
      `preview: ${data.relativePath}`,
    ],
    suggested_next_step: '',
  };
}
